pkg/controller/egress/controller_test.go

```diff
@@ -546,6 +546,76 @@ func TestUpdateEgress(t *testing.T) {
 	checkExternalIPPoolUsed(t, controller, eipFoo2.Name, 0)
 }
 
+// TestRecreateExternalIPPoolWithNewRange tests the case where an ExternalIPPool is deleted, then
+// immediately recreated with a different IP range. Specifically we test the scenario where
+// syncEgress / syncEgressIP are called only once because the DELETE and CREATE events are merged in
+// the workqueue. Ideally, the behavior observed by the user should be the same irrespective of
+// whether the events are merged or not.
+// Note that in an actual cluster, it is very unlikely that both events would be merged.
+func TestRecreateExternalIPPoolWithNewRange(t *testing.T) {
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	eipFoo1 := newExternalIPPool("pool1", "1.1.1.0/24", "", "")
+	egress := &v1beta1.Egress{
+		ObjectMeta: metav1.ObjectMeta{Name: "egressA", UID: "uidA"},
+		Spec: v1beta1.EgressSpec{
+			AppliedTo: v1beta1.AppliedTo{
+				PodSelector: &metav1.LabelSelector{
+					MatchLabels: map[string]string{"app": "foo"},
+				},
+			},
+			EgressIP:       "",
+			ExternalIPPool: eipFoo1.Name,
+		},
+	}
+
+	eventCh := make(chan string, 1)
+	waitForEvent := func(timeout time.Duration) error {
+		select {
+		case <-time.After(timeout):
+			return fmt.Errorf("timeout while waiting for IPPool event")
+		case <-eventCh:
+			return nil
+		}
+	}
+
+	controller := newController(nil, []runtime.Object{eipFoo1, egress})
+	// Register our own event handler to be able to determine when changes have been processed
+	// by the ExternalIPPool controller.
+	controller.externalIPAllocator.AddEventHandler(func(poolName string) {
+		eventCh <- poolName
+	})
+	// A call to RestoreIPAllocations is required for every registered event handler.
+	controller.externalIPAllocator.RestoreIPAllocations(nil)
+	controller.informerFactory.Start(stopCh)
+	controller.crdInformerFactory.Start(stopCh)
+	controller.informerFactory.WaitForCacheSync(stopCh)
+	controller.crdInformerFactory.WaitForCacheSync(stopCh)
+	go controller.externalIPAllocator.Run(stopCh)
+	require.True(t, cache.WaitForCacheSync(stopCh, controller.externalIPAllocator.HasSynced))
+	controller.restoreIPAllocations([]*v1beta1.Egress{egress})
+
+	getEgressIP, _, err := controller.syncEgressIP(egress)
```

Comment:
Could it receive the returned egress and avoid updating the stored `egress` at L613 which may cause race condition?

Reply:
I made the change. By stored `egress`, do you mean the object stored by the fake clientset? They make a copy before adding it to the store:
https://github.com/kubernetes/client-go/blob/c5196ebcc18e1bf29561b7a689fdb121913d221b/testing/fixture.go#L534-L537

but based on the comment, the caller shouldn't modify the object anyway

Reply:
Thanks for the info.